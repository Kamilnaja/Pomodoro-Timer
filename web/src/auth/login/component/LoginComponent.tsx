import { useForm } from "react-hook-form";
import { Login } from "../../../../../types/interfaces";
import { ValidationMessage } from "../../../shared/components/validationMessage/ValidationMessage";
import { AuthState } from "../../store/interfaces/auth.state";
import "./loginComponent.scss";

export interface LoginProps {
  handleSubmit: (data: Login) => void;
  handleClose: () => void;
  auth: AuthState;
}

export const LoginComponent = (props: LoginProps) => {
  const { register, errors, handleSubmit } = useForm();
  const onSubmit = (loginData: Login) => {
    props.handleSubmit(loginData);
  };

  return (
    <div className="login">
      <h2 className="login__header">Please login</h2>
      <form className="login__form form" onSubmit={handleSubmit(onSubmit)}>
        <div className="form__row">
          <label className="form__label">User name</label>
          <input className="form__input" type="text" name="login" required placeholder="John Doe" ref={register({ required: true })} />
          <div className="form__error">{errors.name && <ValidationMessage type={"error"} message={"Name is required"} />}</div>
        </div>
        <div className="form__row">
          <label className="form__label">Password</label>
          <input type="password" name="password" required ref={register({ required: true })} />
          <div className="form__error">{errors.password && <ValidationMessage type={"error"} message={"Password is required"} />}</div>
        </div>
        <button className="form__button" value="WyÅ›lij" type="submit">
          Submit
        </button>
      </form>
      {props.auth.isSuccess && <ValidationMessage message={"success"} type={"success"} />}
    </div>
  );
};
