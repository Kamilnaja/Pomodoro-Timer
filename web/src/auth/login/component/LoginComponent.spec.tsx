import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { shallow, ShallowWrapper } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { act } from 'react-dom/test-utils';
import { createAuthState } from '../../testing/auth.testing.data';
import { LoginComponent } from './LoginComponent';

import React from 'react';
import '@testing-library/jest-dom';

describe('LoginComponent', () => {
  let wrapper: ShallowWrapper;
  const handleSubmit = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<LoginComponent handleSubmit={handleSubmit} auth={createAuthState(true)} />);
  });

  it('Should match snapshot when is success', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('Should match snapshot when form is visible', () => {
    wrapper.setProps({
      auth: createAuthState(false),
    });

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should show validation info at start', () => {
    expect(wrapper.find('.login__success')).toHaveLength(1);

    wrapper.setProps({
      auth: createAuthState(false),
    });

    expect(wrapper.find('.login__success')).toHaveLength(0);
  });

  it('should render form fields', () => {
    render(<LoginComponent handleSubmit={handleSubmit} auth={createAuthState(false, false)} />);

    expect(screen.getByLabelText('Password')).toBeInTheDocument();
    expect(screen.getByLabelText('User login')).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Submit/i })).toBeInTheDocument();
  });

  it('should submit form', async () => {
    wrapper.setProps({
      auth: createAuthState(false, false),
    });
    let getByTestId: any;

    act(() => {
      getByTestId = render(<LoginComponent handleSubmit={handleSubmit} auth={createAuthState(false, false)} />);
    });

    act(() => {
      fireEvent.input(screen.getByLabelText('Password'), {
        target: {
          value: 'Diccus',
        },
      });

      fireEvent.input(screen.getByLabelText('User login'), {
        target: {
          value: 'asdf111111',
        },
      });

      fireEvent.submit(screen.getByRole('button', { name: /Submit/i }));
    });

    expect(shallowToJson(wrapper)).toMatchSnapshot();

    await waitFor(() => {
      expect(handleSubmit).toHaveBeenCalledTimes(1);
    });
  });
});
