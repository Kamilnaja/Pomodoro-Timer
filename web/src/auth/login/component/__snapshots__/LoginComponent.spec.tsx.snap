// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginComponent Should match snapshot when form is visible 1`] = `
<div
  className="login"
>
  <div>
    <form
      className="login__form form"
      onSubmit={[Function]}
    >
      <div
        className="form__row"
      >
        <label
          className="form__label label"
          htmlFor="login"
        >
          User login
        </label>
        <input
          autoFocus={true}
          className="form__input input"
          id="login"
          name="login"
          placeholder="John Doe"
          required={true}
          type="text"
        />
      </div>
      <div
        className="form__row"
      >
        <label
          className="form__label label"
          htmlFor="password"
        >
          Password
        </label>
        <input
          className="form__input input"
          id="password"
          name="password"
          required={true}
          type="password"
        />
      </div>
      <button
        className="form__button"
        type="submit"
        value="WyÅ›lij"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;

exports[`LoginComponent Should match snapshot when is success 1`] = `
<div
  className="login"
>
  <div
    className="login__success"
  >
    <ValidationMessage
      message="ðŸ˜€Login successfull, now close the modal and do some work!"
      type="success"
    />
  </div>
</div>
`;

exports[`LoginComponent should submit form 1`] = `
<div
  className="login"
>
  <div>
    <form
      className="login__form form"
      onSubmit={[Function]}
    >
      <div
        className="form__row"
      >
        <label
          className="form__label label"
          htmlFor="login"
        >
          User login
        </label>
        <input
          autoFocus={true}
          className="form__input input"
          id="login"
          name="login"
          placeholder="John Doe"
          required={true}
          type="text"
        />
      </div>
      <div
        className="form__row"
      >
        <label
          className="form__label label"
          htmlFor="password"
        >
          Password
        </label>
        <input
          className="form__input input"
          id="password"
          name="password"
          required={true}
          type="password"
        />
      </div>
      <button
        className="form__button"
        type="submit"
        value="WyÅ›lij"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
